[FLINK-36379] Improve (Global)Committer with UC disabled

AHeise opened the pull request:
<!--
*Thank you very much for contributing to Apache Flink - we are happy that you want to help us improve Flink. To help the community review your contribution in the best possible way, please go through the checklist below, which will get the contribution into a shape in which it can be best reviewed.*

*Please understand that we do not do this to make contributions to Flink a hassle. In order to uphold a high standard of quality for code contributions, while at the same time managing a large number of contributions, we need contributors to prepare the contributions well, and give reviewers enough contextual information for the review. Please also understand that contributions that do not follow this guide will take longer to review and thus typically be picked up with lower priority by the community.*

## Contribution Checklist

  - Make sure that the pull request corresponds to a [JIRA issue](https://issues.apache.org/jira/projects/FLINK/issues). Exceptions are made for typos in JavaDoc or documentation files, which need no JIRA issue.
  
  - Name the pull request in the form "[FLINK-XXXX] [component] Title of the pull request", where *FLINK-XXXX* should be replaced by the actual issue number. Skip *component* if you are unsure about which is the best component.
  Typo fixes that have no associated JIRA issue should be named following this pattern: `[hotfix] [docs] Fix typo in event time introduction` or `[hotfix] [javadocs] Expand JavaDoc for PuncuatedWatermarkGenerator`.

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.
  
  - Make sure that the change passes the automated tests, i.e., `mvn clean verify` passes. You can set up Azure Pipelines CI to do that following [this guide](https://cwiki.apache.org/confluence/display/FLINK/Azure+Pipelines#AzurePipelines-Tutorial:SettingupAzurePipelinesforaforkoftheFlinkrepository).

  - Each pull request should address only one issue, not mix up code from multiple issues.
  
  - Each commit in the pull request has a meaningful commit message (including the JIRA id)

  - Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.


**(The sections below can be removed for hotfixes of typos)**
-->

## What is the purpose of the change

[FLINK-36287](https://issues.apache.org/jira/browse/FLINK-36287) disabled UC for all inter-sink connections to adhere to the contract of notifyCheckpointCompleted. This allows us to remove some special casing and improve global committer.

## Brief change log

* Refactor sink test assertions 
* Optimize global committers 
* Optimize committers with UC disabled

See commit messages for more details.

## Verifying this change

Covered by tests in
* org.apache.flink.streaming.runtime.operators.sink
* org.apache.flink.streaming.api.connector.sink2
* Various IT cases and E2E tests

I modified / extended the former two suites.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / no / **don't know**)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

fapaul commented on flink-runtime/src/main/java/org/apache/flink/streaming/runtime/translators/GlobalCommitterTransformationTranslator.java:
```diff
@@ -0,0 +1,128 @@
+     * Looks for a committer in the pipeline and aborts on writer. The GlobalCommitter behaves
+     * differently if there is a committer after the writer.
+     */
+    private static boolean hasUpstreamCommitter(DataStream<?> ds) {
```
How safe is this really? Users could have used the global committer with different custom operators in front of that follow either writer or committer semantics.

AHeise replied:
That's a good point. What happens is: if it finds a committer, the new optimized behavior applies. If it doesn't the old unoptimized behavior is used.

So the question is what happens if you have a committer and do stuff afterwards in such a way that the committer suddenly is not doing 2PC anymore until the data reaches the global committer. I haven't been able to find such a scenario. You could have a writer->committer->union->global committer where the union is also directly fed by the writer->union->global committer but even then the output of the committer participated in 2PC.

In short, I'm convinced it's safe that any upstream committer does 2PC already and the global committer can simply piggyback.

Remember that parts of the change is that the 3PC commit protocol that we currently have is is not working properly with final checkpoints. With this fix, we get a working 2PC and "only" continue to violate endInput.

fapaul commented on flink-runtime/src/main/java/org/apache/flink/streaming/runtime/operators/sink/committables/CheckpointCommittableManagerImpl.java:
```diff
@@ -168,13 +169,33 @@ public Collection<CommittableWithLineage<CommT>> commit(Committer<CommT> committ
     }
 
+    /**
+     * Sinks don't use unaligned checkpoints, so we receive all committables of a given upstream
```
This javadoc is now a bit confusing since global committer and committer follow different committing and receival strategies.

fapaul reviewed (commented):
While I understand the idea of the change, I wonder if it the right time or way.

We just cut 2.0 and this is potentially a breaking change since we do not control where users append the global committer in their sink topologies. It's very theoretical and the globalcommitter was marked `@Experimental` but we should at least discuss it.

I am also a bit torn of the change of operation in the `GlobalCommitter` and the usage of the underlying infra structure i.e. `CheckpointCommittableManager`. I probably need another pass to fully understand how well the manager plays with the new semantics of the `GlobalCommitter`.

AHeise commented:
> While I understand the idea of the change, I wonder if it the right time or way.
> 
> We just cut 2.0 and this is potentially a breaking change since we do not control where users append the global committer in their sink topologies. It's very theoretical and the globalcommitter was marked `@Experimental` but we should at least discuss it.
> 
> I am also a bit torn of the change of operation in the `GlobalCommitter` and the usage of the underlying infra structure i.e. `CheckpointCommittableManager`. I probably need another pass to fully understand how well the manager plays with the new semantics of the `GlobalCommitter`.

As discussed offline. The cut was just for 2.0-preview and there is going to be another cut for the actual 2.0.

AHeise commented on flink-runtime/src/main/java/org/apache/flink/streaming/api/connector/sink2/GlobalCommitterOperator.java:
```diff
@@ -47,10 +47,50 @@
+ * error. The state only includes incomplete checkpoints coming from upstream committers not
+ * receiving {@link #notifyCheckpointComplete(long)}. All committables received are successful.
+ *
+ * <p>In rare cases, the GlobalCommitterOperator may be connected to a writer directly. In this
```
Yes, you are right. I'll clarify the comment - the good already acknowledges that.

AHeise commented on flink-runtime/src/main/java/org/apache/flink/streaming/api/connector/sink2/GlobalCommitterOperator.java:
```diff
@@ -127,38 +165,67 @@ public void initializeState(StateInitializationContext context) throws Exception
+        // this is true for the last commit and we need to make sure that all committables are
+        // indeed committed as this function will never be invoked again
+        boolean waitForAllCommitted =
+                lastCompletedCheckpointId == EOI
```
Yes that is a likely scenario. Hence the second part of this statement that will only be true iff there is an EOI committable and it has received all messages.

Then, the only way that could turn into infinite loop is when committables appear out of order (in respect to the checkpointId) which is possible with async retries. However, this applies also to the old version that loops on `endInput` (remember that we need to emit committables post EOI for the final checkpoint)

It will be solved by the next PR that removes async retries.

fapaul reviewed (approved):
Thanks for addressing the comments % please double check that ` org.apache.flink.table.planner.runtime.stream.sql.TableSinkITCase.testSinkDisorderChangeLogWithRank` doesn't fail because of this change 👍

AHeise commented:
> Thanks for addressing the comments % please double check that ` org.apache.flink.table.planner.runtime.stream.sql.TableSinkITCase.testSinkDisorderChangeLogWithRank` doesn't fail because of this change 👍

I double-checked: the test uses the `SinkOperator` wrapping the old `SinkFunction` and has no changed code.

AHeise commented:
@flinkbot run azure

AHeise commented:
@flinkbot run azure

AHeise commented:
Failure is a known issue FLINK-36591.
